casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -21,36 +22,52 @@ object ProtoUtil {
    * c is in the blockchain of b iff c == b or c is in the blockchain of the main parent of b
    */
   // TODO: Move into BlockDAG and remove corresponding param once that is moved over from simulator
-  @tailrec
-  def isInMainChain(internalMap: Map[BlockHash, BlockMessage],
-                    candidate: BlockMessage,
-                    target: BlockMessage): Boolean =
+  def isInMainChain[F[_]: Monad: BlockStore](candidate: BlockMessage,
+                                             target: BlockMessage): F[Boolean] =
     if (candidate == target) {
-      true
+      true.pure[F]
     } else {
-      (for {
-        hdr        <- target.header
-        parentHash <- hdr.parentsHashList.headOption
-        mainParent <- internalMap.get(parentHash)
-      } yield mainParent) match {
-        case Some(parent) => isInMainChain(internalMap, candidate, parent)
-        case None         => false
+      val maybeMainParentHash = ProtoUtil.parents(target).headOption
+      maybeMainParentHash match {
+        case Some(mainParentHash) =>
+          for {
+            mainParent <- BlockStore[F].get(mainParentHash)
+            isInMainChain <- mainParent match {
+                              case Some(parent) => isInMainChain[F](candidate, parent)
+                              case None         => false.pure[F]
+                            }
+          } yield isInMainChain
+        case None => false.pure[F]
       }
     }
 
-  @tailrec
-  def getMainChain(internalMap: Map[BlockHash, BlockMessage],
-                   estimate: BlockMessage,
-                   acc: IndexedSeq[BlockMessage]): IndexedSeq[BlockMessage] = {
+  def getMainChain[F[_]: Monad: BlockStore](
+      estimate: BlockMessage,
+      acc: IndexedSeq[BlockMessage]): F[IndexedSeq[BlockMessage]] = {
     val parentsHashes       = ProtoUtil.parents(estimate)
     val maybeMainParentHash = parentsHashes.headOption
-    maybeMainParentHash flatMap internalMap.get match {
-      case Some(newEstimate) =>
-        getMainChain(internalMap, newEstimate, acc :+ estimate)
-      case None => acc :+ estimate
-    }
+    for {
+      mainChain <- maybeMainParentHash match {
+                    case Some(mainParentHash) =>
+                      for {
+                        updatedEstimate <- getBlock[F](mainParentHash)
+                        mainChain       <- getMainChain[F](updatedEstimate, acc :+ estimate)
+                      } yield mainChain
+                    case None => (acc :+ estimate).pure[F]
+                  }
+    } yield mainChain
   }
 
+  def getBlock[F[_]: Monad: BlockStore](hash: BlockHash): F[BlockMessage] =
```

Comment:
Maybe rename `unsafeGetBlock` to signal the possibility of throwing the error? Alternatively we could maybe try incorporating `MonadError`.

Reply:
Done (unsafeGetBlock).